File: 384-shuffle-an-array.py
import random

class Solution:
    def __init__(self, nums: List[int]):
        self.nums = nums
        self.origin = list(nums)

    def reset(self) -> List[int]:
        self.nums = self.origin
        self.origin = list(self.origin)
        return self.origin

    def shuffle(self) -> List[int]:
        for i in range(len(self.nums)):
            target_idx = random.randrange(i, len(self.nums))
            self.nums[i], self.nums[target_idx] = self.nums[target_idx], self.nums[i]
        return self.nums

File: 384-shuffle-an-array.js
class Solution {
  /**
   * @param {number[]} nums 
   */
  constructor(nums) {
    this.nums = nums;
    this.origin = Array.from(this.nums);
  }

  /**
   * @returns {number[]}
   */
  reset() {
    this.nums = this.origin;
    this.origin = Array.from(this.origin);
    return this.origin;
  }

  /**
   * @returns {number[]}
   */
  shuffle() {
    for (let i = 0; i < this.nums.length; i++) {
      const target = randomRange(i, this.nums.length);
      this._swap(i, target);
    }
    return this.nums;
  }
  
  /**
   * @private
   */
  _swap(leftIdx, rightIdx) {
    const temp = this.nums[leftIdx];
    this.nums[leftIdx] = this.nums[rightIdx];
    this.nums[rightIdx] = temp;
  }
}

function randomRange(min, max) {
  return Math.ceil(Math.random() * (max - min)) + min;
}